### PR TITLE
 Handle C# "for" statement in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Review by mavasani — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs [hunk 1]
```diff
@@ -81,7 +81,23 @@ private BoundForStatement BindForParts(ForStatementSyntax node, Binder originalB
...
+                    if (increment.Kind == BoundKind.StatementList)
+                    {
+                        increment = new BoundBlock(scopeDesignator, locals, ((BoundStatementList)increment).Statements)
+                                            { WasCompilerGenerated = true };
+                    }
+                    else
+                    {
+                        increment = new BoundBlock(increment.Syntax, locals, ImmutableArray.Create(increment))
+                                            { WasCompilerGenerated = true };
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs]: Perhaps move `if` inside `new BoundBlock`:
```
increment = new BoundBlock(
    increment.Syntax,
    locals,
    increment.Kind == BoundKind.StatementList ?
        ((BoundStatementList)increment).Statements :
        ImmutableArray.Create(increment))
    { WasCompilerGenerated = true };
```

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please, need a second review.

**Comment by jcouv** [comment]
Sorry for the delay. Looking now.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -11,30 +11,15 @@ Microsoft.CodeAnalysis.MetadataImportOptions
 Microsoft.CodeAnalysis.MetadataImportOptions.All = 2 -> Microsoft.CodeAnalysis.MetadataImportOptions
 Microsoft.CodeAnalysis.MetadataImportOptions.Internal = 1 -> Microsoft.CodeAnalysis.MetadataImportOptions
 Microsoft.CodeAnalysis.MetadataImportOptions.Public = 0 -> Microsoft.CodeAnalysis.MetadataImportOptions
+Microsoft.CodeAnalysis.OperationKind.CaughtException = 94 -> Microsoft.CodeAnalysis.OperationKind
 Microsoft.CodeAnalysis.OperationKind.ConstructorBodyOperation = 89 -> Microsoft.CodeAnalysis.OperationKind
 Microsoft.CodeAnalysis.OperationKind.Discard = 90 -> Microsoft.CodeAnalysis.OperationKind
-Microsoft.CodeAnalysis.OperationKind.MethodBodyOperation = 88 -> Microsoft.CodeAnalysis.OperationKind
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Not related to this PR: I don't understand why this file always gets a lot of diff noise.
FYI, I filed [https://github.com/dotnet/roslyn/issues/26285](https://github.com/dotnet/roslyn/issues/26285)

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 3]
```diff
@@ -2228,11 +2228,16 @@ internal sealed partial class LazyForEachLoopStatement : BaseForEachLoopStatemen
...
-        public BaseForLoopStatement(ImmutableArray<ILocalSymbol> locals, ILabelSymbol continueLabel, ILabelSymbol exitLabel, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
+        public BaseForLoopStatement(ImmutableArray<ILocalSymbol> locals, ImmutableArray<ILocalSymbol> conditionLocals, 
+            ILabelSymbol continueLabel, ILabelSymbol exitLabel, SemanticModel semanticModel, 
+            SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
             base(LoopKind.For, locals, continueLabel, exitLabel, OperationKind.Loop, semanticModel, syntax, type, constantValue, isImplicit)
         {
+            ConditionLocals = conditionLocals;
         }
 
+        public ImmutableArray<ILocalSymbol> ConditionLocals { get; }
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Perhaps add xml doc

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 4]
```diff
@@ -2606,6 +2611,86 @@ public override IOperation VisitEnd(IEndOperation operation, int? captureIdForRe
...
+            {
+                HandleVariableDeclarations((VariableDeclarationGroupOperation)initialization.Single());
+            }
+            else
+            {
+                VisitStatements(initialization);
+            }
+
+            var start = new BasicBlock(BasicBlockKind.Block); 
+            AppendNewBlock(start);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Nit: extra empty line below

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs [hunk 5]
```diff
@@ -2868,5 +2868,857 @@ static void Main(string[] args)
...
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
+        [Fact]
+        public void ForFlow_04()
+        {
+            string source = @"
+public sealed class MyClass
+{
+    void M(bool result, bool a, bool b, bool c)
+    /*<bind>*/{
+        for (; a ? b : c ;)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: Would it be interesting to test a conditional expression with a throw case?

**Review by jcouv — APPROVED** [review]
LGTM Thanks

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: > Would it be interesting to test a conditional expression with a throw case?

I think not particularly. The purpose of this test is to confirm that condition goes through appropriate conditional branch rewriting. We don't need to test all possible constructs that can cause control flow inside it. Throw expressions are covered by other existing targeted tests.

---
In reply to: [183110773](https://github.com/dotnet/roslyn/pull/26202#discussion_r183110773) [](ancestors = 183110773)

**Pull request merged**